### PR TITLE
rust: 1.77.0

### PR DESCRIPTION
**Pull request by mina86**
Release notes: https://blog.rust-lang.org/2024/03/21/Rust-1.77.0.html

With split_last_chunk bunch of stdx code can now be deleted. \o/

As for other changes:

- Rust now doesn’t consider Clone and Debug derives as using fields of
  a type.  This lead to a handful of dead_code warnings.  I’ve fixed
  them by either removing offending type or adding `allow(dead_code)`.

- Cargo package id format has been stabilised¹ and krates crate
  stopped working.  Why do we care about that crate? It’s a dependency
  of cargo-deny.  To have CI pass I’ve updated taiki-e/install-action
  so that it fetches the newest cargo-deny.

¹ https://github.com/rust-lang/cargo/pull/12914

### CODE REVIEW

File: runtime/near-vm-runner/src/instrument/stack_height/mod.rs [hunk 1]
```diff
@@ -90,6 +90,7 @@ mod thunk;
 ///
 /// This means that the module is invalid.
 #[derive(Debug)]
+#[allow(dead_code)]
```
Comment (nagisa) [inline runtime/near-vm-runner/src/instrument/stack_height/mod.rs:93]: Was this complaining about the field within the struct here (and elsewhere?) Is there a more granular lint we can allow than `dead_code`? Cause this I believe will keep this struct (or enum) around even after it becomes truly unused.

Comment (mina86) [reply runtime/near-vm-runner/src/instrument/stack_height/mod.rs:93]: Complain [here](https://github.com/rust-lang/rust/pull/85200) I guess. ;)

But yeah, I’m not happy with this either.  The best I can think of is to declare a dummy function with `allow(dead_code)` which accesses those fields.

Comment (mina86) [reply runtime/near-vm-runner/src/instrument/stack_height/mod.rs:93]: Turns out types in chain/client/src/adapter.rs weren’t used and Clippy failed to detect that for some reason.  Does derive marks type as used perhaps? In which case `allow(dead_code)` won’t prevent unused type detection.

Comment (nagisa) [reply runtime/near-vm-runner/src/instrument/stack_height/mod.rs:93]: > Turns out types in chain/client/src/adapter.rs weren’t used and Clippy failed to detect that for some reason. Does derive marks type as used perhaps?

Might very well be some of the derives creating a use that's not obvious. e.g. maybe they put the type id into some sort of type registry?

---

> Complain https://github.com/rust-lang/rust/pull/85200 I guess. ;)

Ah, I see. I actually agree with the reasoning in the PR and would remove the field here.

**Review by nagisa — APPROVED** [review]

**Comment by nagisa** [comment]
Looks like the `cargo-deny` check is failing…

**Comment by mina86** [comment]
Yeah, that was next thing on my list. Had to update CI tools. Newer cargo-deny works fine.

**Review by nagisa — APPROVED** [review]

**Pull request merged**